`kinesis_settings must be set when engine_name = "kinesis"` With kinesis_settings Value Set

## Description

I'm hitting
```
│ Error: 1 error occurred:
│       * kinesis_settings must be set when engine_name = "kinesis"
│
│
│
│   with module.replicate_oltp.aws_dms_endpoint.this["rates_kinesis_target"],
│   on .terraform/modules/replicate_oltp/main.tf line 160, in resource "aws_dms_endpoint" "this":
│  160: resource "aws_dms_endpoint" "this" {
```

I have re-initialized my terraform project after cleaning up the `.terraform` directory.

## Versions

- Module version [Required]:
1.5.1

- Terraform version: 
```
terraform -version
Terraform v1.1.9
on linux_amd64
```
- Provider version(s): 
```
# terraform providers

Providers required by configuration:
.
├── provider[registry.terraform.io/shippo/aws] ~> 3.0
├── provider[registry.terraform.io/hashicorp/random]
├── module.dms_write_to_rating_data_sync_policy
│   └── provider[registry.terraform.io/hashicorp/aws] >= 3.35.0
├── module.postgres_dms_instance_access
│   └── provider[registry.terraform.io/hashicorp/aws] >= 3.0.0
├── module.replicate_oltp
│   └── provider[registry.terraform.io/hashicorp/aws] >= 4.6.0
├── module.replication_backup
│   └── provider[registry.terraform.io/shippo/aws] ~> 3.0
└── module.defaults
    ├── provider[registry.terraform.io/hashicorp/aws]
    └── provider[terraform.io/builtin/terraform]
```

## Reproduction Code [Required]
```
  endpoints = {
    rates_kinesis_target = {
      endpoint_id   = "${local.name}-rates-kinesis-target"
      endpoint_type = "target"
      engine_name   = "kinesis"

      kinesis_settings = {
        # These options are described in
        # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Target.Kinesis.html
        service_access_role_arn        = aws_iam_role.dms_write_to_rating_data_sync.arn
        stream_arn                     = data.aws_kinesis_stream.rating_replication_stream.arn
        partition_include_schema_table = true
        include_partition_value        = true
      }
    }
```

Steps to reproduce the behavior:

Are you using workspaces?: Yes
Have you cleared the local cache (see Notice section above)?: Yes
<!-- List steps in order that led up to the issue you encountered -->
```
Terraform init -backend-config=envs/dev-main/backend.tf
Terraform plan
```

## Expected behavior

Terraform should be able to find the `kinesis_settings` as described in the documentation.

## Actual behavior

`kinesis_settings` is not being discovered in the plan step.

### Terminal Output Screenshot(s)

![Screen Shot 2022-06-21 at 5 35 17 PM](https://user-images.githubusercontent.com/88306949/174918709-35ec18ef-8e45-4523-8a7d-df97e778415d.png)


## Additional context

<!-- Add any other context about the problem here -->

# Trace Logs From Terraform Plan
```
2022-06-22T23:50:40.484Z [WARN]  Provider "registry.terraform.io/hashicorp/aws" produced an invalid plan for module.replicate_oltp.aws_dms_endpoint.this["postgresql_source"], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .tags: planned value cty.NullVal(cty.Map(cty.String)) does not match config value cty.MapValEmpty(cty.String)
      - .redshift_settings: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
2022-06-22T23:50:40.485Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for module.replicate_oltp.aws_dms_endpoint.this["postgresql_source"]
2022-06-22T23:50:40.485Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for module.replicate_oltp.aws_dms_endpoint.this["postgresql_source"]
2022-06-22T23:50:40.485Z [ERROR] vertex "module.replicate_oltp.aws_dms_endpoint.this[\"rates_kinesis_target\"]" error: 1 error occurred:
        * kinesis_settings must be set when engine_name = "kinesis"
2022-06-22T23:50:40.485Z [TRACE] vertex "module.replicate_oltp.aws_dms_endpoint.this[\"rates_kinesis_target\"]": visit complete, with errors
2022-06-22T23:50:40.485Z [TRACE] writeChange: recorded Create change for module.replicate_oltp.aws_dms_endpoint.this["postgresql_source"]
2022-06-22T23:50:40.486Z [TRACE] vertex "module.replicate_oltp.aws_dms_endpoint.this[\"postgresql_source\"]": visit complete
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "root" errored, so skipping
2022-06-22T23:50:40.485Z [TRACE] provider.terraform-provider-aws_v4.19.0_x5: Served request: @caller=github.com/hashicorp/terraform-plugin-go@v0.9.1/tfprotov5/tf5server/server.go:791 @module=sdk.proto tf_proto_version=5.2 tf_provider_addr=provider tf_resource_type=aws_dms_endpoint tf_req_id=4fa6f44f-c38e-fbf5-37af-3b50cbc67640 tf_rpc=PlanResourceChange timesta
mp=2022-06-22T23:50:40.485Z
2022-06-22T23:50:40.486Z [TRACE] vertex "module.replicate_oltp.aws_dms_endpoint.this": dynamic subgraph encountered errors: 1 error occurred:
        * kinesis_settings must be set when engine_name = "kinesis"
2022-06-22T23:50:40.486Z [ERROR] vertex "module.replicate_oltp.aws_dms_endpoint.this" error: 1 error occurred:
        * kinesis_settings must be set when engine_name = "kinesis"
2022-06-22T23:50:40.486Z [TRACE] vertex "module.replicate_oltp.aws_dms_endpoint.this": visit complete, with errors
2022-06-22T23:50:40.486Z [TRACE] vertex "module.replicate_oltp.aws_dms_endpoint.this (expand)": dynamic subgraph encountered errors: 1 error occurred:
        * kinesis_settings must be set when engine_name = "kinesis"
2022-06-22T23:50:40.486Z [ERROR] vertex "module.replicate_oltp.aws_dms_endpoint.this (expand)" error: 1 error occurred:
        * kinesis_settings must be set when engine_name = "kinesis"
2022-06-22T23:50:40.486Z [TRACE] vertex "module.replicate_oltp.aws_dms_endpoint.this (expand)": visit complete, with errors
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "module.replicate_oltp.output.endpoints (expand)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "module.replicate_oltp.aws_dms_replication_task.this (expand)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "module.replicate_oltp.output.replication_tasks (expand)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "module.replicate_oltp.aws_dms_event_subscription.this (expand)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "module.replicate_oltp.output.event_subscriptions (expand)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "module.replicate_oltp (close)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/aws\"] (close)" errored, so skipping
2022-06-22T23:50:40.486Z [TRACE] dag/walk: upstream of "root" errored, so skipping
2022-06-22T23:50:40.486Z [INFO]  backend/local: plan operation completed
╷
│ Error: 1 error occurred:
│       * kinesis_settings must be set when engine_name = "kinesis"
│
│
│
│   with module.replicate_oltp.aws_dms_endpoint.this["rates_kinesis_target"],
│   on .terraform/modules/replicate_oltp/main.tf line 160, in resource "aws_dms_endpoint" "this":
│  160: resource "aws_dms_endpoint" "this" {
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/antonbabenko/pre-commit-terraform
-    rev: v1.68.1
+    rev: v1.72.2
     hooks:
       - id: terraform_fmt
       - id: terraform_validate
@@ -23,7 +23,7 @@ repos:
           - '--args=--only=terraform_standard_module_structure'
           - '--args=--only=terraform_workspace_remote'
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v4.2.0
+    rev: v4.3.0
     hooks:
       - id: check-merge-conflict
       - id: end-of-file-fixer

--- a/examples/complete/README.md
+++ b/examples/complete/README.md
@@ -17,8 +17,6 @@ To run this example you need to execute:
 ```bash
 $ terraform init
 $ terraform plan
-# due to https://github.com/hashicorp/terraform/issues/4149
-$ terraform apply --target module.msk_cluster
 $ terraform apply
 ```
 

--- a/main.tf
+++ b/main.tf
@@ -1,7 +1,8 @@
 locals {
   subnet_group_id = var.create && var.create_repl_subnet_group ? aws_dms_replication_subnet_group.this[0].id : var.repl_instance_subnet_group_id
 
-  partition = data.aws_partition.current.partition
+  partition  = data.aws_partition.current.partition
+  dns_suffix = data.aws_partition.current.dns_suffix
 }
 
 data "aws_partition" "current" {}
@@ -19,7 +20,7 @@ data "aws_iam_policy_document" "dms_assume_role" {
     actions = ["sts:AssumeRole"]
 
     principals {
-      identifiers = ["dms.amazonaws.com"]
+      identifiers = ["dms.${local.dns_suffix}"]
       type        = "Service"
     }
   }
@@ -34,7 +35,7 @@ data "aws_iam_policy_document" "dms_assume_role_redshift" {
     actions = ["sts:AssumeRole"]
 
     principals {
-      identifiers = ["redshift.amazonaws.com"]
+      identifiers = ["redshift.${local.dns_suffix}"]
       type        = "Service"
     }
   }
@@ -176,7 +177,8 @@ resource "aws_dms_endpoint" "this" {
 
   # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Target.Elasticsearch.html
   dynamic "elasticsearch_settings" {
-    for_each = try([each.value.elasticsearch_settings], [])
+    for_each = length(lookup(each.value, "elasticsearch_settings", {})) == 0 ? [] : [each.value.elasticsearch_settings]
+
     content {
       endpoint_uri               = elasticsearch_settings.value.endpoint_uri
       error_retry_duration       = lookup(elasticsearch_settings.value, "error_retry_duration", null)
@@ -187,7 +189,8 @@ resource "aws_dms_endpoint" "this" {
 
   # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Target.Kafka.html
   dynamic "kafka_settings" {
-    for_each = try([each.value.kafka_settings], [])
+    for_each = length(lookup(each.value, "kafka_settings", {})) == 0 ? [] : [each.value.kafka_settings]
+
     content {
       broker                         = kafka_settings.value.broker
       include_control_details        = lookup(kafka_settings.value, "include_control_details", null)
@@ -212,7 +215,8 @@ resource "aws_dms_endpoint" "this" {
 
   # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Target.Kinesis.html
   dynamic "kinesis_settings" {
-    for_each = try([each.value.kinesis_settings], [])
+    for_each = length(lookup(each.value, "kinesis_settings", {})) == 0 ? [] : [each.value.kinesis_settings]
+
     content {
       include_control_details        = lookup(kinesis_settings.value, "include_control_details", null)
       include_null_and_empty         = lookup(kinesis_settings.value, "include_null_and_empty", null)
@@ -228,7 +232,8 @@ resource "aws_dms_endpoint" "this" {
 
   # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Source.MongoDB.html
   dynamic "mongodb_settings" {
-    for_each = can(each.value["mongodb_settings"]) ? [each.value.mongodb_settings] : []
+    for_each = length(lookup(each.value, "mongodb_settings", {})) == 0 ? [] : [each.value.mongodb_settings]
+
     content {
       auth_mechanism      = lookup(mongodb_settings.value, "auth_mechanism", null)
       auth_source         = lookup(mongodb_settings.value, "auth_source", null)
@@ -242,7 +247,8 @@ resource "aws_dms_endpoint" "this" {
   # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Source.S3.html
   # https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Target.S3.html
   dynamic "s3_settings" {
-    for_each = can(each.value["s3_settings"]) ? [each.value.s3_settings] : []
+    for_each = length(lookup(each.value, "s3_settings", {})) == 0 ? [] : [each.value.s3_settings]
+
     content {
       add_column_name                   = lookup(s3_settings.value, "add_column_name", null)
       bucket_folder                     = lookup(s3_settings.value, "bucket_folder", null)

